daemon/rrd-update.c

```diff
@@ -243,34 +247,69 @@ void rb_rrd_update(rb_poller *poll)
                 log_errorx("couldn't create directory for raw file: %s : %s",
                             path,  strerror(errno));
                 free(parent);
-                /* Try the next item */
-                continue;
+                break; /* next raw file */
             }
             free(parent);
-        
-            fp = fopen(path, "a");
-            if(fp == NULL)
-            {
-                log_errorx("couldn't open raw file: %s for writing : %s",
-                            path,  strerror(errno));
-                /* Try the next item */
-                continue;
+
+            fd = open(path, O_WRONLY|O_APPEND|O_CREAT);
+            if(fd == -1) {
+                log_errorx("open raw file for append failed: %s: %s", path, strerror(errno));
+                break; /* next raw file */
             }
 
-            /* Write item record to raw file */
-            fprintf(fp, "%" PRId64 "\t", time);
-            if(item->reference)
-                fprintf(fp, "%s", item->reference);
-            else
-                fprintf(fp, "%s", item->field);
-            fprintf(fp, "\t");
-            if(item->vtype == VALUE_REAL)
-                fprintf(fp, "%" PRId64, item->v.i_value);
-            else if(item->vtype == VALUE_FLOAT)
-                fprintf(fp, "%.4lf", item->v.f_value);
-            fprintf(fp, "\n");
-
-            fclose(fp);
+	    write_sample(fd, path, &time, item);
+
+	    if (close(fd) == -1) {
+		    log_errorx("close of raw file failed: %s: %s", path, strerror(errno));
+	    }
         }
     }
 }
+
+static void
+write_sample(int fd, const char* path, const time_t *time, const rb_item *item)
```

Comment:
Why is write_sample() getting the path?  It is not used anywhere... and it already gets an open fd

Reply:
This seems to have been missed - I'd suggest dropping the path argument from write_sample (if it already takes the fd).